core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -286,58 +291,117 @@ protected void doClose() throws IOException {
         IOUtils.close(masterFD, nodesFD);
     }
 
-    @Override
-    public DiscoveryNode localNode() {
-        return clusterService.localNode();
-    }
-
-    @Override
-    public String nodeDescription() {
-        return clusterName.value() + "/" + clusterService.localNode().getId();
-    }
-
     /** start of {@link PingContextProvider } implementation */
     @Override
     public DiscoveryNodes nodes() {
-        return clusterService.state().nodes();
+        return clusterState().nodes();
     }
 
     @Override
     public ClusterState clusterState() {
```

Comment:
I'm a bit on the fence with this one. In case of no master, this state has NO_MASTER_BLOCK and is definitely not committed. Another aspect is that the public interface method which is implemented here (from the `Discovery` interface) has as only purpose to expose the state to `MasterService` to calculate the next state that is to be committed. Whereas currently this state is committed (except if NO_MASTER_BLOCK in place), a future discovery implementation would expose a state here, however, that's not committed yet, but just the best state found among the nodes that is used as a basis by `MasterService` to calculate the next state that is to be committed.